notifications_utils/polygons.py

```diff
@@ -233,6 +250,23 @@ def estimated_area(self):
             polygon.area for polygon in self
         ) * self.square_degrees_to_square_miles
 
+    def ratio_of_intersection_with(self, polygons):
+        '''
+        Given another Polygons object, this works how much the two
+        overlap, as a fraction of the area of this Polygons object.
+        '''
+        if self.estimated_area == 0:
+            return 0
+        return sum(
```

Comment:
Is there a risk of double-counting here if multiple polygons overlap the same area? For example, if someone selections "England" + "Hackney Marshes", then a subsequent overlap would count their shared area twice.

It looks like the intended use of this is to calculate the overlap of a drawn polygon over bounding boxes at the same level in an R Tree (from a quick skim of the PR linked to this one), so there doesn't seem to be a risk in that scenario?

However, since this is meant to be a generic utility, it's worth making the non-multi-overlapping assumption explicit in the method description.

Reply:
Yeah, this is a risk more generally so I’ve added a note in 524c45d

## For custom polygons 

At some point we should add validation for user-created polygons. At the moment we have no guards against:
- polygons which overlap each other
- polygons which self-intesect (hourglass or bow tie shaped polygons)

We can check with the Flood Warning team to see what kind of problems they check for in their data set and apply the same checks.

# For our library

Nothing stops double-counting at the moment – here’s Hackney alone:

![image](https://user-images.githubusercontent.com/355079/113280540-55825f80-92dc-11eb-82ec-fcb697f9544e.png)

And here it is with some of its wards

![image](https://user-images.githubusercontent.com/355079/113280510-4b606100-92dc-11eb-8c03-7232d7f5ea89.png)

Reply:
Oh dear 😬. Thanks for adding the warning, and it's good to know it's a general problem - then we can work towards a general fix for it.